File: lib/utils/getIntraNodeCrossingsUsingCircle.ts
import { NodeWithPortPoints } from "lib/types/high-density-types"

type Chord = {
  endpoints: [number, number]
  connectionName: string
  rootConnectionName: string
}

/**
 * Maps a boundary point to a 1D perimeter coordinate.
 * Starting at top-left corner, going clockwise:
 * - Top edge (y=ymax): t = x - xmin
 * - Right edge (x=xmax): t = W + (ymax - y)
 * - Bottom edge (y=ymin): t = 2W + H + (xmax - x)
 * - Left edge (x=xmin): t = 2W + 2H + (y - ymin)
 */
function perimeterT(
  p: { x: number; y: number },
  xmin: number,
  xmax: number,
  ymin: number,
  ymax: number,
): number {
  const W = xmax - xmin
  const H = ymax - ymin
  const eps = 1e-6

  // Top edge
  if (Math.abs(p.y - ymax) < eps) {
    return p.x - xmin
  }
  // Right edge
  if (Math.abs(p.x - xmax) < eps) {
    return W + (ymax - p.y)
  }
  // Bottom edge
  if (Math.abs(p.y - ymin) < eps) {
    return W + H + (xmax - p.x)
  }
  // Left edge
  if (Math.abs(p.x - xmin) < eps) {
    return 2 * W + H + (p.y - ymin)
  }

  // Point is not exactly on boundary - find closest edge
  const distTop = Math.abs(p.y - ymax)
  const distRight = Math.abs(p.x - xmax)
  const distBottom = Math.abs(p.y - ymin)
  const distLeft = Math.abs(p.x - xmin)

  const minDist = Math.min(distTop, distRight, distBottom, distLeft)

  if (minDist === distTop) {
    return Math.max(0, Math.min(W, p.x - xmin))
  }
  if (minDist === distRight) {
    return W + Math.max(0, Math.min(H, ymax - p.y))
  }
  if (minDist === distBottom) {
    return W + H + Math.max(0, Math.min(W, xmax - p.x))
  }
  // Left edge
  return 2 * W + H + Math.max(0, Math.min(H, p.y - ymin))
}

/**
 * Check if two perimeter coordinates are coincident (within epsilon)
 */
function areCoincident(t1: number, t2: number, eps: number = 1e-6): boolean {
  return Math.abs(t1 - t2) < eps
}

/**
 * Count necessary crossings between chords on a circle using the interleaving criterion.
 * Two chords (a,b) and (c,d) with a < b and c < d cross iff: a < c < b < d OR c < a < d < b
 *
 * Chords that share a coincident endpoint do NOT count as crossing.
 *
 * Uses O(n^2) algorithm to correctly handle coincident endpoints.
 */
function countChordCrossings(chords: Chord[]): {
  total: number
  differentRoot: number
} {
  if (chords.length < 2) {
    return { total: 0, differentRoot: 0 }
  }

  // Normalize each chord so first endpoint is smaller
  const normalizedChords = chords.map((chord) => {
    const [t1, t2] = chord.endpoints
    const endpoints =
      t1 < t2 ? ([t1, t2] as [number, number]) : ([t2, t1] as [number, number])
    return { ...chord, endpoints }
  })

  let crossings = 0
  let differentRootCrossings = 0

  // Check all pairs of chords
  for (let i = 0; i < normalizedChords.length; i++) {
    const chord1 = normalizedChords[i]
    const [a, b] = chord1.endpoints
    for (let j = i + 1; j < normalizedChords.length; j++) {
      const chord2 = normalizedChords[j]
      const [c, d] = chord2.endpoints

      // Skip if chords share a coincident endpoint
      if (
        areCoincident(a, c) ||
        areCoincident(a, d) ||
        areCoincident(b, c) ||
        areCoincident(b, d)
      ) {
        continue
      }

      // Two chords cross iff their endpoints interleave: a < c < b < d OR c < a < d < b
      if ((a < c && c < b && b < d) || (c < a && a < d && d < b)) {
        crossings++
        if (chord1.rootConnectionName !== chord2.rootConnectionName) {
          differentRootCrossings++
        }
      }
    }
  }

  return { total: crossings, differentRoot: differentRootCrossings }
}

/**
 * Compute intra-node crossings using the circle/perimeter mapping approach.
 *
 * This is topologically correct: two connections MUST cross if their boundary
 * points interleave around the perimeter, regardless of which side of the
 * rectangle they are on.
 *
 * Returns the same output structure as getIntraNodeCrossings.
 */
export const getIntraNodeCrossingsUsingCircle = (node: NodeWithPortPoints) => {
  const xmin = node.center.x - node.width / 2
  const xmax = node.center.x + node.width / 2
  const ymin = node.center.y - node.height / 2
  const ymax = node.center.y + node.height / 2

  // Group port points by connectionName
  const connectionPointsMap = new Map<
    string,
    {
      rootConnectionName: string
      points: Array<{ x: number; y: number; z: number }>
    }
  >()

  for (const pp of node.portPoints) {
    const existing = connectionPointsMap.get(pp.connectionName) ?? {
      rootConnectionName: pp.rootConnectionName ?? pp.connectionName,
      points: [],
    }
    // Avoid duplicate points
    if (
      !existing.points.some((p) => p.x === pp.x && p.y === pp.y && p.z === pp.z)
    ) {
      existing.points.push({ x: pp.x, y: pp.y, z: pp.z })
    }
    connectionPointsMap.set(pp.connectionName, existing)
  }

  // Separate same-layer pairs from transition pairs
  const sameLayerPairsByZ = new Map<number, Chord[]>()
  const transitionPairs: Chord[] = []
  let numEntryExitLayerChanges = 0

  for (const [connectionName, group] of connectionPointsMap) {
    const { points, rootConnectionName } = group
    if (points.length < 2) continue

    // Get the two endpoints for this connection
    const p1 = points[0]
    const p2 = points[1]

    // Map to perimeter coordinates
    const t1 = perimeterT(p1, xmin, xmax, ymin, ymax)
    const t2 = perimeterT(p2, xmin, xmax, ymin, ymax)

    if (p1.z === p2.z) {
      // Same layer - add to the layer's chord list
      const z = p1.z
      const chords = sameLayerPairsByZ.get(z) ?? []
      chords.push({
        endpoints: [t1, t2],
        connectionName,
        rootConnectionName,
      })
      sameLayerPairsByZ.set(z, chords)
    } else {
      // Transition pair - different layers
      numEntryExitLayerChanges++
      transitionPairs.push({
        endpoints: [t1, t2],
        connectionName,
        rootConnectionName,
      })
    }
  }

  // Count same-layer crossings (per layer, then sum)
  let numSameLayerCrossings = 0
  let numDifferentRootSameLayerCrossings = 0
  for (const [z, chords] of sameLayerPairsByZ) {
    const crossingCounts = countChordCrossings(chords)
    numSameLayerCrossings += crossingCounts.total
    numDifferentRootSameLayerCrossings += crossingCounts.differentRoot
  }

  // Count transition pair crossings
  // Transition pairs can cross each other regardless of layer
  const transitionCrossingCounts = countChordCrossings(transitionPairs)

  return {
    numSameLayerCrossings,
    numDifferentRootSameLayerCrossings,
    numEntryExitLayerChanges,
    numTransitionPairCrossings: transitionCrossingCounts.total,
  }
}

File: rust/autorouter-bindings/src/ported/utils/get_intra_node_crossings_using_circle.rs
use indexmap::IndexMap;
use serde_json::{Value, json};
use autorouting_drc::math_utils::{js_min, js_max};

fn perimeter_t(p: &Value, xmin: f64, xmax: f64, ymin: f64, ymax: f64) -> f64 {
    let x = p["x"].as_f64().expect("Point x required"); let y = p["y"].as_f64().expect("Point y required");
    let w = xmax - xmin; let h = ymax - ymin; let eps = 1e-6;
    if (y - ymax).abs() < eps { return x - xmin; }
    if (x - xmax).abs() < eps { return w + (ymax - y); }
    if (y - ymin).abs() < eps { return w + h + (xmax - x); }
    if (x - xmin).abs() < eps { return 2.0 * w + h + (y - ymin); }
    let top = (y - ymax).abs(); let right = (x - xmax).abs(); let bottom = (y - ymin).abs(); let left = (x - xmin).abs();
    let min = js_min(js_min(js_min(top, right), bottom), left);
    if min == top { return js_max(0.0, js_min(w, x - xmin)); }
    if min == right { return w + js_max(0.0, js_min(h, ymax - y)); }
    if min == bottom { return w + h + js_max(0.0, js_min(w, xmax - x)); }
    2.0 * w + h + js_max(0.0, js_min(h, y - ymin))
}

fn count_chord_crossings(chords: &[[f64; 2]]) -> usize {
    if chords.len() < 2 { return 0; }
    let normalized: Vec<[f64; 2]> = chords.iter().map(|&[a,b]| if a < b { [a,b] } else { [b,a] }).collect();
    let mut crossings = 0;
    for (i, &[a,b]) in normalized.iter().enumerate() {
        for &[c,d] in &normalized[i + 1..] {
            if (a-c).abs() < 1e-6 || (a-d).abs() < 1e-6 || (b-c).abs() < 1e-6 || (b-d).abs() < 1e-6 { continue; }
            if (a < c && c < b && b < d) || (c < a && a < d && d < b) { crossings += 1; }
        }
    }
    crossings
}

pub fn get_intra_node_crossings_using_circle(node: &Value) -> Value {
    let cx = node["center"]["x"].as_f64().expect("Center x required"); let cy = node["center"]["y"].as_f64().expect("Center y required");
    let width = node["width"].as_f64().expect("Width required"); let height = node["height"].as_f64().expect("Height required");
    let xmin = cx - width / 2.0; let xmax = cx + width / 2.0; let ymin = cy - height / 2.0; let ymax = cy + height / 2.0;
    let mut groups: IndexMap<&str, Vec<&Value>> = IndexMap::new();
    for p in node["portPoints"].as_array().expect("Ports required") {
        let points = groups.entry(p["connectionName"].as_str().expect("Connection name required")).or_default();
        if !points.iter().any(|other| other["x"] == p["x"] && other["y"] == p["y"] && other["z"] == p["z"]) { points.push(p); }
    }
    let mut same_layers: Vec<(Value, Vec<[f64; 2]>)> = Vec::new(); let mut transitions = Vec::new(); let mut layer_changes = 0;
    for points in groups.values() {
        if points.len() < 2 { continue; }
        let a = points[0]; let b = points[1];
        let chord = [perimeter_t(a,xmin,xmax,ymin,ymax), perimeter_t(b,xmin,xmax,ymin,ymax)];
        if a["z"] == b["z"] {
            if let Some((_, chords)) = same_layers.iter_mut().find(|(z,_)| z == &a["z"]) { chords.push(chord); }
            else { same_layers.push((a["z"].clone(), vec![chord])); }
        } else { layer_changes += 1; transitions.push(chord); }
    }
    json!({"numSameLayerCrossings":same_layers.iter().map(|(_, chords)| count_chord_crossings(chords)).sum::<usize>(),
        "numEntryExitLayerChanges":layer_changes,"numTransitionPairCrossings":count_chord_crossings(&transitions)})
}
